### PR TITLE
Python: decide decorator forms and placeholder bodies from the source

### PR DESCRIPTION
**Pull request by graemerocher**
> Stacked on #13224 (part 20/32 of the Python fix stack; merge in order — GitHub retargets each PR when its base merges).

Three heuristics of the Python compiler (`micronaut-inject-python`) guessed at the meaning of a Python source instead of reading it, and each guess broke a documented use of the framework. The three fixes share the same principle: the syntactic form of the source decides.

### Placeholder method bodies made a concrete bean abstract

A method whose body is only `...` was always compiled as abstract, so a `@Singleton` (a Pub/Sub listener, say) whose handlers were written as placeholders became an abstract class that is not a bean candidate: "Could not find the bean to execute the method". The module ports had to spell empty bodies as `pass`.

A placeholder body is now abstract only where Python itself would not instantiate the class or where Micronaut implements the method: in a class extending `ABC` or `Protocol`, when the method is decorated with `@abstractmethod`, or in a class carrying an `@Introduction` stereotype (`@Client`, a Data repository, an AI service), whose methods are implemented by the introduction advice. In any other class it is a method returning `None`. The processor records the placeholder as a new `FunctionDef.hasPlaceholderBody` fact next to `isAbstract` (which it now sets for `ABC` and `Protocol` members and `@abstractmethod` only); `PythonMethodElement.isAbstract()` and `AbstractPythonClassElement.isAbstract()` add the introduction rule, which needs the class stereotypes the Java element model knows. This keeps the interface rules of #13216 intact: an `ABC`/`Protocol` with placeholder bodies is still abstract and still compiles to an interface, and an introduction type with placeholder bodies is still an introduction interface.

### A positional class-valued decorator argument was taken for bare decoration

The generated decorators decided at runtime whether they were applied bare (`@Singleton`) or called (`@Singleton(...)`) by asking whether their single argument was callable and unbound in the caller's frame (`_getframe`). A class-valued annotation value that is not a module-level name, `@TestMethodOrder(MethodOrderer.OrderAnnotation)`, met the test, was returned as the "target", and Python then instantiated the foreign class with the test class: `TypeError: invalid instantiation of foreign object` at import, which poisoned the whole package. Only the keyword form `value=...` worked.

The decision is now syntactic and made at compile time, as for a Java annotation: the transformer rewrites a bare `@X` to `@X()` for every generated decorator, in the compile-time tree (already the case) *and* in the runtime tree, and a call expression is always a factory invocation. The generated decorators are plain factories with no runtime heuristic, so every argument is an annotation value. The runtime transformer learns the decorator names from the same imports the compile-time transformer generates decorators for (`from jakarta.inject import Singleton`, `import jakarta.inject as i`, `from jakarta import inject`, star imports).

### Custom annotation functions could not be applied bare

A custom annotation is a function returning a decorator (`@FilterBinding def AdultMales(): def decorator(target): ...; return decorator`). Bare `@AdultMales` called it with the class, `TypeError: AdultMales() takes 0 positional arguments but 1 was given`, while `@AdultMales()` worked and bare `AdultMales` inside `Annotated[...]` worked.

The transformer now applies the same rewrite to custom annotation functions: an `AnnotationFunctionScanner` finds, before a module is transformed, the names it binds to annotation functions, as the processor recognises them (a top-level function of the decorator factory shape decorated with `micronaut_annotation`, with a Java annotation applicable to annotation types, or with another annotation function), following imports to the other modules of the same source root (`from .AdultMales import AdultMales`, `from filters.AdultMales import AdultMales`). For that the transform stage is given each source's package and source root, as the parse stage already is. A function whose single required positional parameter is the decorated target is a plain Python decorator and is left alone whatever it returns — `@Around def Timed(target): return target` as well as the wrapping shape `def Traced(func): def wrapper(...): ...; return wrapper` — it is applied bare by construction. The runtime transformer records the annotations of a Java package imported as a module (`import jakarta.inject as i`, `from jakarta import inject`) under the name the package is bound to and matches `@i.Singleton` on that qualifier, so an unrelated local decorator of the same simple name is not rewritten. The scanner caches the modules it follows for one scan only, not across compilations.

### Reproducers

`PlaceholderMethodBodySpec` (a `@Singleton` with only `...` bodies is resolvable and its `@Executable` methods run; `ABC`/`Protocol`/`@abstractmethod` placeholders stay abstract; an introduction type's placeholders are implemented by the advice) and `DecoratorFormSpec` (`@MethodOrder(Orderers.Alphabetical)` with a `Class<?> value()` member on a class and a method, recorded and importable; a custom annotation applied bare and with parentheses on classes and methods, in the defining module and in importing modules with absolute and relative imports; a wrapping decorator `@Traced` on a bean method returning `traced:ok`), plus unit tests of the transformer and processor sources (the wrapping shape, the qualified package decorators, the per-scan cache). All new scenarios fail on `5.2.x`.

The `interfaces.adoc` of #13216 describes a placeholder body as always abstract; once both are merged that sentence should read "written with a `...` placeholder body in an `ABC`, a `Protocol` or an introduction type". A new guide page, "Annotations as Decorators", documents the decorator forms and placeholder bodies.

### Module tests that can be re-enabled

* micronaut-gcp #1543: `@PubSubListener` snippets can keep their `...` bodies instead of `pass` (the `DISABLED_TESTS.md` migration rule about placeholder bodies no longer applies).
* micronaut-test #1533: `@TestMethodOrder(MethodOrderer.OrderAnnotation)` works positionally (`TestOutcomeListenerTest`, `PreDestroyTest`).
* micronaut-coherence #1064: `@AdultMales` can be applied bare on `AdultMalesFilterFactory`.

Includes the `closeClassLoader` move in the Java compiler driver that the checkstyle `InnerTypeLast` rule requires of every change to these sources.

### CODE REVIEW

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py [hunk 1]
```diff
@@ -55,27 +56,177 @@ def unresolved_java_io_import_error(name: str, kind: str = 'import') -> str:
...
+
+    def scan(self, module: ast.Module) -> Set[str]:
+        for statement in module.body:
+            if isinstance(statement, ast.ImportFrom):
+                self._scan_import(statement)
+            elif isinstance(statement, ast.FunctionDef) and self.declares_annotation(statement):
+                self.annotation_functions.add(statement.name)
+        return self.annotation_functions
+
+    def declares_annotation(self, function: ast.FunctionDef) -> bool:
```
Comment (dstepanov) [inline inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py:178]: **Blocking regression:** `returns_nested_function` + an annotation-applicable annotation is treated as a factory without looking at the parameters, so a plain wrapping decorator applied bare gets rewritten to a call:

```python
@Bindable
def Traced(func):
    def wrapper(*a, **k): return "traced:" + func(*a, **k)
    return wrapper

# on a bean method
@Traced
def hello(self): return "ok"
```

On #13224 this returns `traced:ok`; here the context fails to start with `TypeError: Traced() missing 1 required positional argument: 'func'`. The new docs say a function taking the target directly is left alone, but nothing checks that. Suggest not rewriting functions with exactly one required positional parameter, and adding this case to `DecoratorFormSpec`.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py [hunk 2]
```diff
@@ -1260,6 +1425,18 @@ def visit_Import(self, node: ast.Import):
         ]
         return ast.copy_location(ast.Import(names=names), node)
 
+    def _track_package_decorators(self, java_module: str):
```
Comment (dstepanov) [inline inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: Nit: a package/star import adds the simple name of every annotation in that Java package to `generated_decorators`, and the lookup ignores the qualifier — so an unrelated local decorator with the same simple name also gets turned into a call. Match on the qualified/bound name instead.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py [hunk 3]
```diff
@@ -55,27 +56,177 @@ def unresolved_java_io_import_error(name: str, kind: str = 'import') -> str:
...
+class AnnotationFunctionScanner:
+    """
+    Finds the names a module binds to custom annotation functions: top-level functions of the decorator
+    factory shape (``def Ann(...): def decorator(target): ...; return decorator``) that are decorated with
+    ``micronaut_annotation``, with a Java annotation applicable to annotation types, or with another custom
+    annotation function, as the processor recognises them. Imported modules of the same source root are
+    scanned in turn, so a bare ``@Ann`` can be told from a factory call ``@Ann(...)`` wherever ``Ann`` is used.
+    """
+
+    _cache: Dict[str, Set[str]] = {}
```
Comment (dstepanov) [inline inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: Nit: this class-level cache is keyed on the scanned file's own mtime/size only; changes to modules it imports from, or to the classpath, aren't considered, so a long-lived compiler context (incremental builds) can reuse a stale answer.

**Review by dstepanov — CHANGES_REQUESTED** [review]
Placeholder bodies and abstractness look right (ABC/Protocol stay abstract, introduction types and interceptors handled). The decorator-form detection, though, turns the most common Python decorator shape into a factory call — a regression versus #13224, reproduced locally. Details inline.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py (hunk 1 above)
Comment (graemerocher) [reply inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py:178]: Done in 61bafa6015: `takes_target_directly` (exactly one required positional parameter) excludes a function from `declares_annotation`, whatever it returns, so a wrapping decorator stays bare in the runtime and compile-time trees. Your case is `DecoratorFormSpec."test a wrapping decorator taking the target is applied bare"` (`traced:ok`; the context failed to start before) and `test_wrapping_decorator_returning_a_nested_function_is_not_an_annotation_factory` in the transformer unit tests; the guide states the rule.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py (hunk 2 above)
Comment (graemerocher) [reply inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: Done: the runtime transformer records the annotations of a package import under the name the package is bound to (`package_decorators`, `i` for `import jakarta.inject as i`, `jakarta.inject` for the unaliased form, `inject` for `from jakarta import inject`) and `_is_annotation_decorator` matches an attribute decorator on that qualifier only; a bare local `Executable` next to `import ... as annotation` is left alone (`test_package_import_decorators_are_matched_by_their_qualifier`). A star import still binds the simple names, as Python does.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py (hunk 3 above)
Comment (graemerocher) [reply inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: Done: the cache is now per scan (created by the top-level scanner and shared with the scanners of the modules it follows, so a circular import is still cut), keyed on the path; nothing is kept across module transforms or compilations. `test_imported_module_scan_is_not_cached_across_scans` rewrites an imported module with the same size and mtime and sees the new answer.

**Comment by graemerocher** [comment]
Review feedback addressed in 61bafa6015 (one commit on top, no rebase):

* **Wrapping decorators stay bare** (blocking regression): `takes_target_directly` — exactly one required positional parameter — excludes a function from the annotation-function scan whatever it returns, so `@Bindable def Traced(func): ... return wrapper` applied as `@Traced` is left alone in both the compile-time and the runtime tree. Reproducer: `DecoratorFormSpec."test a wrapping decorator taking the target is applied bare"` (`traced:ok`; before: `TypeError: Traced() missing 1 required positional argument: 'func'` at context start) and `test_wrapping_decorator_returning_a_nested_function_is_not_an_annotation_factory`. The decorators guide states the rule.
* **Package imports match on the qualifier**: the runtime transformer records the annotations of a package import under the bound name (`i`, `jakarta.inject`, `inject`) in `package_decorators` and `_is_annotation_decorator` matches attribute decorators on that qualifier only; a local `Executable` next to `import ... as annotation` stays bare. A star import still binds the simple names, as Python does. Tests: `test_package_import_decorators_are_matched_by_their_qualifier`, `test_unaliased_package_import_decorators_are_matched_by_the_package_name`.
* **Scanner cache per scan**: created by the top-level scanner and shared with the scanners of the modules it follows (circular imports still cut), keyed on the path; nothing survives a module transform. `test_imported_module_scan_is_not_cached_across_scans` rewrites an imported module with the same size and mtime and sees the new answer.

All three unit tests and the spec fail on the previous head. Verified with `-Ppython-ci`: `:micronaut-inject-python:test` (Python unit tests included), `:micronaut-inject-python-test:test` (full), checkstyle clean.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py [hunk 4]
```diff
@@ -55,27 +56,206 @@ def unresolved_java_io_import_error(name: str, kind: str = 'import') -> str:
...
+    def scan(self, module: ast.Module) -> Set[str]:
+        for statement in module.body:
+            if isinstance(statement, ast.ImportFrom):
+                self._scan_import(statement)
+            elif isinstance(statement, ast.FunctionDef) and self.declares_annotation(statement):
+                self.annotation_functions.add(statement.name)
+        return self.annotation_functions
+
+    def declares_annotation(self, function: ast.FunctionDef) -> bool:
+        if takes_target_directly(function) or not returns_nested_function(function):
```
Comment (dstepanov) [inline inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: New regression from the fix: a custom annotation function whose only required parameter is a *member*, e.g. `@Bindable def Tagged(value: str)`, now matches `takes_target_directly` and drops out of `annotation_functions`. A custom annotation decorated with it (`@Tagged("x") def Composite(): ...`) is then no longer recognised either, so a bare `@Composite` on a bean isn't rewritten and startup fails with `TypeError: Composite() takes 0 positional arguments but 1 was given` (passed on `05a4ec8b74`). Suggest applying the exclusion only when the function doesn't have the factory shape, or using the parameter check only to decide bare vs called, not whether something is an annotation.

Nit: a `def Star(*args)` wrapper has no required positional parameter, so it's still rewritten to `Star()` and fails with `IndexError` (worked on #13224).

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py [hunk 5]
```diff
@@ -1228,11 +1422,18 @@ def visit_ImportFrom(self, node: ast.ImportFrom):
 
         java_module = self._to_java_import_module(node.module)
         for alias in node.names:
+            variable_name = alias.asname if alias.asname else alias.name
             if alias.name == '*':
+                # the annotations of the package are bound to their own names
+                self.generated_decorators.update(self._package_annotation_names(java_module))
```
Comment (dstepanov) [inline inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: A star import still adds every annotation's simple name here, but in Python a later local `def Primary(target)` shadows the star-imported name. `from micronaut.context.annotation import *` plus a local `def Primary(target)` applied as `@Primary` fails at startup with `TypeError: Primary() missing 1 required positional argument: 'target'`; it worked on #13224. Dropping the name when the module defines (or assigns) it after the import would match Python's binding.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py [hunk 6]
```diff
@@ -1260,6 +1465,33 @@ def visit_Import(self, node: ast.Import):
...
+            for class_element in self.callback_get_class_elements(java_module) or []
+            if self._is_annotation_class(class_element)
+        }
+
+    def _is_annotation_decorator(self, decorator) -> bool:
+        if isinstance(decorator, ast.Attribute):
+            qualifier = dotted_name(decorator.value)
+            if qualifier in self.package_decorators:
+                return decorator.attr in self.package_decorators[qualifier]
+        return super()._is_annotation_decorator(decorator)
```
Comment (dstepanov) [inline inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py:1565]: Nit: this fallback still matches on the attribute name alone, so `@other.Executable` is rewritten whenever `Executable` is also imported by name.

**Review by dstepanov — CHANGES_REQUESTED** [review]
Thanks — the blocking case is fixed: a bare `@Traced` with `def Traced(func)` and `def Traced(func, *, prefix="x")` work again, factories (`@Prefixed("p")`) and dual-use decorators (`@Dual` / `@Dual()`) still work, qualified imports are matched on the qualifier, and the scan cache is per scan. Checked with a local probe on this head, the previous head and #13224 (`-Ppython-ci`); `DecoratorFormSpec`, `PlaceholderMethodBodySpec`, `:micronaut-inject-python:test` (125) and checkstyle pass.

Still open, details inline: a new regression for custom annotations with a required member, and a star-imported name still wins over a local decorator of the same name (worked on #13224).

Not answered yet — generated decorators applied bare outside the transformer: `Bean = Singleton` then `@Bean class Aliased` leaves `Aliased` a function with no error (it stayed a class on #13224). Raising a clear error when the single argument is a class/function and the annotation has no required members would make that visible.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py (hunk 4 above)
Comment (graemerocher) [reply inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: Done in cd65483a49, your second suggestion: the parameter check only decides bare vs called. The scanner keeps `Tagged` an annotation function (it maps each name to whether a bare application passes the target), so `@Tagged("x") def Composite()` is recognised and a bare `@Composite` is rewritten again, while a bare `@Tagged`/`@Traced` is left alone. `def Star(*args)` (only `*args`, no positional parameter) is treated the same and stays bare. Spec: "a custom annotation with a required member is a stereotype of another custom annotation" (failed with your `TypeError` before); unit tests `test_custom_annotation_with_a_required_member_stays_an_annotation_function`, `test_wrapping_decorator_with_only_varargs_is_applied_bare`.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py (hunk 5 above)
Comment (graemerocher) [reply inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: Done: the runtime transformer collects the names the module defines or assigns itself (`locally_bound_names`) and a star import binds the package annotations minus those, so a local `def Primary(target)` applied as `@Primary` is left alone (spec "a local decorator shadows a star-imported annotation...", `Primary() missing 1 required positional argument` before; unit test `test_a_local_decorator_shadows_a_star_imported_annotation`).

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py (hunk 6 above)
Comment (graemerocher) [reply inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py:1565]: Done: an attribute decorator is matched on its qualifier (a package import) or as a nested annotation of a generated decorator bound by name (`@Outer.Inner`), never on the attribute name alone: `@other.Executable` stays as written with `Executable` imported by name (`test_an_attribute_decorator_is_not_matched_on_its_attribute_name_alone`).

**Comment by graemerocher** [comment]
Second round, cd65483a49:

* **Required member vs target** (regression): the parameter check now only decides bare vs called. `AnnotationFunctionScanner` maps each annotation function to whether a bare application passes it the target (exactly one required positional parameter, or only `*args`); such a function is left bare, the others are rewritten. `@Tagged("x") def Composite()` is recognised again and a bare `@Composite` works; `def Star(*args)` stays bare. Spec: "a custom annotation with a required member is a stereotype of another custom annotation".
* **Star import shadowing**: names the module defines or assigns itself are removed from what a star import binds; `def Primary(target)` applied as `@Primary` next to `from micronaut.context.annotation import *` works again (spec "a local decorator shadows a star-imported annotation and an alias of a generated decorator is applied bare").
* **Attribute decorators** are matched on the qualifier or as `@Outer.Inner` of a generated decorator, never on the attribute name alone.
* **Bare application outside the transformer**: `Bean = Singleton` in the same module is tracked (the alias is the same factory, so `@Bean` becomes `@Bean()` and `Aliased` stays a class, covered by the same spec). For what the transformer cannot see (an alias in another module, `getattr`), a generated factory whose single positional argument is a class or a function now raises `TypeError: @Singleton was applied bare to <class ...> through a name the compiler did not recognise as the annotation; write @Singleton() or apply it under the name it is imported as` — unless the annotation's `value` member holds a class (`@TestMethodOrder(MethodOrderer.OrderAnnotation)`, `@Replaces(Orderer)` are untouched; a nested annotation value, the inner `decorator` of another factory, is not a target). `PythonAnnotationTypes.valueMemberTypeName` answers the member type; the guide documents the error.

All new unit tests (6) and spec cases (2) fail on 61bafa6015. Verified with `-Ppython-ci`: `:micronaut-inject-python:test` (Python unit tests included), `:micronaut-inject-python-test:test` (full), checkstyle clean.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py [hunk 7]
```diff
@@ -55,27 +56,214 @@ def unresolved_java_io_import_error(name: str, kind: str = 'import') -> str:
...
+    """
+    Whether a bare application of the function passes it the decorated target: it has exactly one required
+    positional parameter (``def Traced(func): ...``, or a member without a default, ``def Tagged(value)``), or
+    only ``*args`` (``def Star(*args): ...``). Such a function is never rewritten to a call, whatever it
+    returns: ``@Traced`` stays ``Traced(target)``.
+    """
+    arguments = function.args
+    positional = list(arguments.posonlyargs) + list(arguments.args)
+    if not positional:
+        return arguments.vararg is not None
```
Comment (dstepanov) [inline inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py:113]: Nit: the flip side of the `Star(*args)` fix — a varargs-only *factory* (`@Bindable def Tags(*values: str): def decorator(target): ...; return decorator`) applied bare as `@Tags` now receives the class and silently replaces it with `decorator` (`isClass=false`; it stayed a class on `61bafa6015`). The guide states the rule, and the new bare-use guard only exists in generated decorators, so a custom function gets no error. Fine if varargs-only factories aren't a supported shape.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py [hunk 8]
```diff
@@ -1228,11 +1459,18 @@ def visit_ImportFrom(self, node: ast.ImportFrom):
 
         java_module = self._to_java_import_module(node.module)
         for alias in node.names:
+            variable_name = alias.asname if alias.asname else alias.name
             if alias.name == '*':
+                # the annotations of the package are bound to their own names, unless the module binds a name itself
+                self.generated_decorators.update(self._package_annotation_names(java_module) - self.locally_bound_names)
```
Comment (dstepanov) [inline inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: Nit: `locally_bound_names` (1412) counts local definitions wherever they are, so `def Primary(target)` *before* `from micronaut.context.annotation import *` still shadows the star-imported name, although in Python the star import rebinds it. `@Primary class X` then fails with the new `TypeError: @Primary was applied bare…` (worked on `61bafa6015` and #13224). Unusual order and a clear error; counting only names bound after the star import would match Python.

**Review by dstepanov — APPROVED** [review]
Thanks, all addressed. Verified on `cd65483a49` with `-Ppython-ci` against the previous head and #13224: a bare `@Composite` built on `@Tagged("x")` works on classes and methods; a local `def Primary(target)` after a star import receives the class; `Bean = Singleton` in the same module keeps `Aliased` a class, and unseen bare uses (`getattr`, cross-module alias, `Named(Plain)`) now raise a clear `TypeError` while `@Replaces(Orig)` and `@Requirements(Requires(...))` still work; `def Star(*args)` and `@other.Executable` are fine; `Traced`, `Prefixed("p")`, `@Dual`/`@Dual()` and qualified imports keep working. `DecoratorFormSpec`, `PlaceholderMethodBodySpec`, `:micronaut-inject-python:test` (125) and checkstyle pass. Two low-severity edge cases inline.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py (hunk 7 above)
Comment (graemerocher) [reply inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py:113]: Documented rather than changed (ca4cb85381): varargs-only annotation functions are not a supported bare shape — the compiler treats `def Tags(*values)` like any function taking only `*args` (it receives the target when applied bare), so `decorators.adoc` now says such an annotation is written `@Tags(...)`/`@Tags()` and that a bare application hands it the target; an annotation meant to be applied bare should declare a regular member with a default. A fail-fast guard would need the transformer to distinguish a varargs-only factory from a plain `def Star(*args)` decorator, which have the same signature.

File: inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py (hunk 8 above)
Comment (graemerocher) [reply inject-python/src/main/resources/GRAALPY-VFS/io.micronaut/micronaut-inject-python/src/micronaut_transformer.py]: Fixed in ca4cb85381: the shadowing set of a star import is now the set of names the module binds *after* that import (`names_bound_after_star_import`, computed per star-import statement in `visit_Module`; a star import nested in a block keeps the whole-module set as before), so `def Primary(target)` before `from micronaut.context.annotation import *` is rebound by the import as in Python. New `DecoratorFormSpec` case: `@Primary` on a class after that order is the annotation (`hasDeclaredAnnotation(Primary)`, no `primary` attribute).

**Comment by graemerocher** [comment]
Follow-up pushed (ca4cb85381): a star import now only yields to names the module binds after it (a `def Primary(target)` before the import is rebound, as in Python) with a `DecoratorFormSpec` case, and `decorators.adoc` documents that a varargs-only annotation function applied bare receives the target (not a supported bare shape). `DecoratorFormSpec`, `PlaceholderMethodBodySpec`, `:micronaut-inject-python:test` (125) and checkstyle pass locally; the full `:micronaut-inject-python-test:test -Ppython-ci` is running and I will follow up only if it turns something up.

**Review by dstepanov — APPROVED** [review]
Thanks — both nits handled. The star-import shadowing set is now the names bound *after* the import, which matches Python, and documenting varargs-only annotation functions as a called-only shape is fine given a plain `def Star(*args)` decorator has the same signature. Verified on `ca4cb85381` with `-Ppython-ci`: `DecoratorFormSpec` + `PlaceholderMethodBodySpec` (11) and `:micronaut-inject-python:test` (125, incl. the transformer unit tests) pass, checkstyle clean.

**Review by dstepanov — APPROVED** [review]
Re-approving `3880a8b290`: rebase onto the updated parent, commit bodies unchanged (checked by diffing the old and new patches).

**Review by dstepanov — APPROVED** [review]
Re-approving `97b0c0361c`: rebase onto the updated parent, all commit bodies unchanged (diffed the old and new patches).

**Review by dstepanov — APPROVED** [review]
Re-approving `27f25fca7b`: rebase onto the updated parent, all commit bodies unchanged (diffed the old and new patches).

**Review by dstepanov — APPROVED** [review]
Re-approving `2bc542513f`: rebase onto the updated parent, all commit bodies unchanged (diffed the old and new patches).

**Review by dstepanov — APPROVED** [review]
Re-approving `a591f6fdfd`: rebase onto the updated parent, all six commit bodies unchanged (`git range-diff` against `2bc542513f`).

---
_Generated by [Claude Code](https://claude.ai/code)_
